ENH,BUG: Timeseries downsampling can now deal with MaskedColumn

mhvk opened the pull request:
Also fixes a bug introduced in main, where MaskedQuantity can no longer be downsampled (I did not label it bug to avoid the auto-checker going nuts...  it is "affects-dev" anyway)

The changelog entry only reflects the improvement, since the bug that is fixed has not yet hit any release. It does mention that column types are now preserved -- which at least is true for the column types that I added, but test coverage is generally a bit sub-par, so I am not sure it covers everything. But at least it should be quite a bit better than it was.

Fixes #17875

cc @prajwel, since this changes the code introduced in #17541 to be able to deal with masks -- which was not really tested before at all, but more or less worked

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

mhvk commented:
Failures are unrelated, though at least one seems real (see #18024); will wait with re-running until there has been review.

orionlee commented:
The PR passed the sample codes with TESS fits LC in https://github.com/astropy/astropy/issues/17991#issue-2981109992.

---
Some comments on the test `test_downsample_with_masked_array()`: 
1. the second sub-test involving `down2` and `expected2` is not really exercised.
2. It seems that `assert_array_equal()` does not quite do what we need when masks are involved.

### For `down2` and `expected2`

- `down2` and `expected2` are not asserted
- `down2` is not created properly because the line `m.mask[3] = True` does not change the `ts` instance used for downsampling.
- consider to create a 3rd sub-test, asserting a masked element is treated the same as an `np.nan` for the purpose of downsampling

The test will be: 
```python
@pytest.mark.parametrize("masked_cls", [MaskedColumn, Masked(u.Quantity)])
def test_downsample_with_masked_array(masked_cls):
    # See gh-17991
    m = masked_cls([0.0, 1.0, -np.inf, 3.0, 4.0], mask=False)
    m.mask[2] = True
    ts = TimeSeries(time=INPUT_TIME, data=[m], names=["m"])
    assert isinstance(ts["m"], masked_cls)
    down = aggregate_downsample(
        ts, time_bin_start=INPUT_TIME[:-1:2], time_bin_end=INPUT_TIME[1::2]
    )
    expected = masked_cls([0.5, 3.0], mask=False)
    assert_array_equal(down["m"], expected)

    m.mask[3] = True
    ts2 = TimeSeries(time=INPUT_TIME, data=[m], names=["m"])
    down2 = aggregate_downsample(
        ts2, time_bin_start=INPUT_TIME[:-1:2], time_bin_end=INPUT_TIME[1::2]
    )
    expected2 = masked_cls([0.5, 3.0], mask=[False, True])
    assert_array_equal(down2["m"], expected2)

    # a masked element and a np.nan are treated the same way
    m.mask[3] = False
    m[3] = np.nan
    ts3 = TimeSeries(time=INPUT_TIME, data=[m], names=["m"])
    down3 = aggregate_downsample(
        ts3, time_bin_start=INPUT_TIME[:-1:2], time_bin_end=INPUT_TIME[1::2]
    )
    expected3 = masked_cls([0.5, 3.0], mask=[False, True])
    assert_array_equal(down3["m"], expected3)
```

### `assert_array_equal()` with masked elements

It seems that `assert_array_equal()` does not quite do what we need in the test.
When an element is masked, the assertion always passes (for the elemnt)

E.g., in the assertion for `expected2` of `[0.5 --]`, `assert_array_equal()` does not really assert the value at index 1. Hypothetically, if the actual value is `[0.5, 9.9]`, the assertion would still pass, while we would want it to fail for the test.

```python
# it passes
assert_array_equal(
    Column([0.5, 9.9]),  # hypothetical actual
    MaskedColumn([0.5, 3.0], mask=[False, True]),  # expected2
)

```

orionlee commented on astropy/timeseries/tests/test_downsample.py:
```diff
@@ -314,6 +345,26 @@ def test_downsample_edge_cases(time, time_bin_start, time_bin_end):
+    down2 = aggregate_downsample(
+        ts, time_bin_start=INPUT_TIME[:-1:2], time_bin_end=INPUT_TIME[1::2]
+    )
+    expected2 = masked_cls([0.5, 3.0], mask=[False, True])
```
See https://github.com/astropy/astropy/pull/18023#issuecomment-2816361077  (hard to write inline)

orionlee commented on docs/changes/timeseries/18023.feature.rst:
```diff
@@ -0,0 +1,3 @@
+Downsampling now works correctly also on ``MaskedColumn`` with possibly masked
```
Should ``Masked`` be stated as well?

mhvk replied:
Indeed, done.

mhvk commented:
Thanks, @orionlee! You're right that the test was incorrect -- I adapted your solution and added a little function to test equality of masked classes (which probably should live elsewhere, but that seems out of scope...). Of course, that brought out a deficiency too, so that's now fixed as well...

orionlee commented on astropy/timeseries/tests/test_downsample.py:
```diff
@@ -27,6 +29,14 @@
 )
 
 
+def assert_masked_equal(a, b):
```
It might be out of scope. This helper is similar (but functionally not identical) to:

https://github.com/astropy/astropy/blob/72a1d42bd395caf07bc4ff1b1ba7312c99fdbfd3/astropy/utils/masked/tests/test_masked.py#L22-L26

mhvk replied:
I probably should unify that one and the one introduced here into `astropy.tests`. I think it is a bit out of scope here, but am happy to be convinced otherwise.

astrofrog reviewed (approved):
This looks good to me, thank you for working on this @mhvk! Are you happy for me to merge this?
